### ISSUE TITLE
Add peers protocol matching mechanism

### ISSUE THREAD

**Issue by y0sher** [spacemeshos/go-spacemesh#87]
#### Feature description
When a session is established between two nodes they should be able to know which protocols each node is supporting right after handshake. 
If there's no matching protocols drop that node for now.

#### Main Reason for adding this feature
The other scenario is trying to communicate with a node, then getting unsupported protocol msg, then trying other protocol or query protocols.
this can be saved and be easier with that feature.

**Comment by lucasalcantara** [comment]
Is it anyone working on this issue? if not I'd like to work on it.

**Comment by y0sher** [comment]
This issue is quite complicated since our existing protocols aren't modular enough to do protocol matching yet. 
No one working on this issue though its spec isn't mature enough to start working on , please feel free to drop in architecture proposals for this feature. 
I will try to put in some more info about what we have later. (Everything is in P2P package for now of course).

So, you can take it, but architecture and planning should be done first before implementation.

**Comment by lucasalcantara** [comment]
so I think is better go to another issue.

**Comment by y0sher** [comment]
this can function as capabilities or services matching. eventually it can help to discover poets or make sure all nodes use the same version.

**Comment by lrettig** [comment]
One additional thought from #2122 (a duplicate issue):

> this raises the question, what exactly defines a protocol version? We may need to manually hardcode this each time it changes - but at the very least, it should change when message or serialization format change, and probably when consensus rules change too. (Nodes with incompatible config flags should also not try to talk to one another. This would be easier to compare programmatically, if we're not already doing that.)

As @y0sher [pointed out](https://github.com/spacemeshos/go-spacemesh/issues/2122#issuecomment-693336462), we do currently need to manually update ClientVersion and we haven't yet tried to release a backwards compatible version (although in practice multiple releases/builds are sometimes backwards compatible).

It's also interesting to consider our plans for network ID (https://github.com/spacemeshos/SMIPS/issues/26) and including something like a hash of the config params - in practice we need to match both the protocol build and these params across peers.

**Comment by lrettig** [comment]
also related: #1918
